### PR TITLE
kloud/provider: add Azure support

### PR DESCRIPTION
**Pull request by rjeczalik**
~~deps: https://github.com/koding/koding/pull/9182~~

### CODE REVIEW

**Comment by rjeczalik** [comment]
ping

**Comment by cihangir** [comment]
could you separate the sdk from this PR into another one? github ui does not show the diff

**Comment by rjeczalik** [comment]
@cihangir Yup sorry, actually I forgot to split it - here it is: https://github.com/koding/koding/pull/9309

**Comment by rjeczalik** [comment]
split and rebased

File: go/src/koding/kites/kloud/provider/azure/azure.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+	if cred, ok := m.Credential.Credential.(*Cred); ok {
+		meta.Location = cred.Location
+	}
+
+	return meta
+}
+
+type PublishData struct {
+	Subscriptions []Subscription `xml:"PublishProfile>Subscription"`
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/azure.go]: xml :/

File: go/src/koding/kites/kloud/provider/azure/azure.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	// settings file contains only one subscription, we default to it.
+	if meta.SubscriptionID == "" {
+		if n := len(pb.Subscriptions); n != 1 {
+			return fmt.Errorf("publish settings contain %d subscriptions, please specify which one to use", n)
+		}
+
+		return nil
+	}
+
+	var found bool
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/azure.go]: you could simplify with

```
  for _, sub := range pb.Subscriptions {
      if sub.ID == meta.SubscriptionID {
          return nil
      }
  }

  return errors.New("specified subscription ID does not exist")
```

Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/azure/azure.go]: yup, will do - when I coded it originally I had two more checks after pb.Subscriptions that's why no early return here

File: go/src/koding/kites/kloud/provider/azure/azure.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	"fmt"
+
+	"koding/kites/kloud/stack"
+	"koding/kites/kloud/stack/provider"
+
+	"github.com/Azure/azure-sdk-for-go/management"
+	vm "github.com/Azure/azure-sdk-for-go/management/virtualmachine"
+)
+
+var p = &provider.Provider{
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/azure.go]: you dont need to have this in global scope AFAIS

File: go/src/koding/kites/kloud/provider/azure/bootstrap_test.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"encoding/json"
+	"testing"
+
+	"koding/kites/kloud/provider/azure"
+)
+
+func TestBootstrapJSONTmpl(t *testing.T) {
+	cases := []*azure.BootstrapConfig{
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/bootstrap_test.go]: could you please add test cases for other variables in this struct as well?

File: go/src/koding/kites/kloud/provider/azure/bootstrap_test.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+	for i, cas := range cases {
+		tmpl, err := azure.NewBootstrapTmpl(cas)
+		if err != nil {
+			t.Fatalf("%d: NewBootstrapTmpl()=%s", i, err)
+		}
+
+		var v map[string]interface{}
+
+		if err = json.Unmarshal(tmpl, &v); err != nil {
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/bootstrap_test.go]: you can have expected results in a table and check them here

File: go/src/koding/kites/kloud/provider/azure/machine.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+func (m *Machine) Bootstrap() *Bootstrap {
+	return m.BaseMachine.Bootstrap.(*Bootstrap)
+}
+
+func (m *Machine) Meta() *Meta {
+	return m.BaseMachine.Metadata.(*Meta)
+}
+
+func (m *Machine) Start(ctx context.Context) (interface{}, error) {
+	id, err := m.AzureVMClient.StartRole(m.Meta().HostedServiceID, m.Meta().InstanceID, m.Meta().InstanceID)
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/machine.go]: could we do something like for all the functions?
```
	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
	defer cancel()

	closeChan := make(chan struct{})
	go func() {
		<-ctx.Done()
		close(closeChan)
	}()
	WaitForOperation(closeChan)
```

File: go/src/koding/kites/kloud/provider/azure/machine.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+	}
+}
+
+// isNotFound tests whether err is a resource not found Azure's error.
+func isNotFound(err error) bool {
+	// No idea why they return everything by a value - checking
+	// for both in case this would get fixed.
+	e, ok := err.(*management.AzureError)
+	if !ok {
+		ee, ok := err.(management.AzureError)
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/machine.go]: srsly? sometimes with pointer sometimes not?

File: go/src/koding/kites/kloud/provider/azure/stack.go [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+	"koding/kites/kloud/stack/provider"
+	"koding/kites/kloud/userdata"
+)
+
+//go:generate $GOPATH/bin/go-bindata -mode 420 -modtime 1470666525 -pkg azure -o bootstrap.json.tmpl.go bootstrap.json.tmpl
+//go:generate go fmt bootstrap.json.tmpl.go
+
+var tmpl = template.Must(template.New("").Parse(mustAsset("bootstrap.json.tmpl")))
+
+type BootstrapConfig struct {
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/stack.go]: linting issues

File: go/src/koding/kites/kloud/provider/azure/stack.go [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+	return nil
+}
+
+func (s *Stack) BootstrapTemplates(c *stack.Credential) ([]*stack.Template, error) {
+	boot := c.Bootstrap.(*Bootstrap)
+
+	cfg := &BootstrapConfig{
+		TeamSlug:           s.BootstrapArg().GroupName,
+		HostedServiceName:  "koding-hs-" + c.Identifier,
+		StorageServiceName: strings.ToLower("kodings" + c.Identifier),
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/stack.go]: s/kodings/koding-s- ?

File: go/src/koding/kites/kloud/provider/azure/stack.go [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+		return nil, err
+	}
+
+	return []*stack.Template{
+		{Content: string(bootstrapTmpl)},
+		{Content: string(ruleTmpl)},
+	}, nil
+}
+
+func (s *Stack) ApplyTemplate(c *stack.Credential) (*stack.Template, error) {
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/stack.go]: could you please separate this function into smaller ones?

File: go/src/koding/kites/kloud/provider/azure/stack_test.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    SchemaVersion="2.0"
+    PublishMethod="AzureServiceManagementAPI">
+    <Subscription
+      ServiceManagementUrl="https://management.core.windows.net"
+      Id="aa3ed1eb-ff8c-4312-b2f9-11cf611563b4"
+      Name="Pay-As-You-Go"
+      ManagementCertificate="..." />
+    <Subscription
+      ServiceManagementUrl="https://management.core.windows.net"
+      Id="8a523531-2d66-469d-9228-718ad2a58eff"
```
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/stack_test.go]: where these ids come from? safe to be public?

File: go/src/koding/kites/kloud/provider/azure/stack.go (hunk 10 above)
Comment (cihangir) [inline go/src/koding/kites/kloud/provider/azure/stack.go]: could you write test?

**Review by cihangir — CHANGES_REQUESTED** [review]

File: go/src/koding/kites/kloud/provider/azure/machine.go (hunk 7 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/azure/machine.go]: Currently they use struct values - without pointers, I'm handling also pointers if they had a second thought and fixed this eventually.

File: go/src/koding/kites/kloud/provider/azure/bootstrap_test.go (hunk 5 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/azure/bootstrap_test.go]: I don't test for expected values, only for valid JSON - so for every case throwaway `v` variable is used.

File: go/src/koding/kites/kloud/provider/azure/stack.go (hunk 9 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/azure/stack.go]: Storage name can't have `-` characters, that's why it differs.

File: go/src/koding/kites/kloud/provider/azure/stack_test.go (hunk 11 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/azure/stack_test.go]: Yes, safe, they are not valid. If they are, then not ours ;-)

File: go/src/koding/kites/kloud/provider/azure/bootstrap_test.go (hunk 4 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/azure/bootstrap_test.go]: Done.

**Comment by rjeczalik** [comment]
@cihangir All addressed, ptal.

File: go/src/koding/kites/kloud/provider/azure/stack.go [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+
+	if err := tmpl.Execute(&buf, cfg); err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+func mustAsset(s string) string {
+	p, err := Asset(s)
```
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/azure/stack.go]: You can use `MustAsset` generated by go-bindata.

**Review by cihangir — APPROVED** [review]
we can merge after deployment

**Review by ppknap — APPROVED** [review]

File: go/src/koding/kites/kloud/provider/azure/stack.go [hunk 13]
```diff
@@ -250,7 +254,7 @@ func (s *Stack) injectEndpointRules(vm map[string]interface{}) {
 		endpoints = append(endpoints, vmSSH)
 	}
 
-	vm["endpoint"] = endpoints
+	vm["endpoint"] = provider.ToSlice(endpoints)
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/azure/stack.go]: /cc @ppknap 

```
--- PASS: TestAzure_ApplyTemplate (0.01s)
    --- PASS: TestAzure_ApplyTemplate/basic_stack (0.00s)
    --- PASS: TestAzure_ApplyTemplate/custom_stack (0.00s)
```

Comment (ppknap) [inline go/src/koding/kites/kloud/provider/azure/stack.go]: Should be good for now. However, this method does not work when shadowed element is inside slice.

**Comment by rjeczalik** [comment]
What about we merge azure before deploy? The feature would be a ~nop without client support.

**Comment by gokmen** [comment]
@cihangir we can merge this since it seems it won't affect deploy afais.

**Pull request merged**